### ISSUE TITLE
Keyboard shortcut to stage/unstage file in SCM view

### ISSUE THREAD

**Issue by kgrossjo** [Microsoft/vscode#24916]
- VSCode Version: 1.11.2
- OS Version: macOS Sierra 10.12.4

Steps to Reproduce:

1. Open a project under git, with some changed files
2. Ctrl-Shift-G to open the SCM view.  It has a commit message at the top and a list of Changes below.
3. Tab to go to the list of Changes.
4. Cursor down until an interesting file is selected.

Now I can do Return to see a diff.  After that, the file is highlighted and there are icons in the SCM view for reverting this file (curved arrow) or for staging it (plus sign).

But I can't find a keyboard shortcut for the "stage" icon.  I tried Return, Ctrl-Return, Space, Ctrl-Space, S, Shift-S, the Plus key, Cmd-Down (because that opens a file in Explorer view), Cmd-Up, Cmd-Right, Cmd-Left.

I also tried Cmd-Shift-P to find the command, but it doesn't have a keyboard shortcut printed.  I also tried to right-click the file, which gave me a context menu (with "stage" in it), again without keyboard shortcut.

**Comment by joaomoreno** [comment]
Try <kbd>Shift F10</kbd> and then pick the desired action.

**Comment by kgrossjo** [comment]
Hm.  `F10` invokes a macOS action (to show all windows of the application), and `Shift-F10` does the same but with slow motion animation.

But it provided a starting point: In the keyboard shortcuts I searched for `F10`, found the `Shift-F10` command, and reassigned it to `cmd-F10`.  So now when I'm in a text editor, I can use it and I get a nice context menu.

But the keyboard shortcut has "When `editorTextFocus`", and so it doesn't work in the SCM view.

I also searched the reference documentation to see if there is a "When" setting (I think it's called "context"?) that I can use to trigger when the SCM view has focus.  But I've not been able to find it.

The issue was closed, but for me, it's not resolved yet.  @joaomoreno is it okay with you to reopen this ticket?

**Comment by joaomoreno** [comment]
Absolutely.

**Comment by kgrossjo** [comment]
Fn-F10 mutes the loudspeakers.  Fn-Shift-F10 seems to do the same.

By now, it's starting to look as if these actions aren't possible via the keyboard.  There is an action for "git add" that I can activate via cmd-shift-P, but that one seems to apply to the currently focused editor, not to the file that's selected in the SCM view.

Thanks for your help, @joaomoreno I really appreciate it.  I wish I could be more positive in my messages.

**Comment by joaomoreno** [comment]
No worries @kgrossjo. We should definitely get this story better.

**Comment by cleidigh** [comment]
@kgrossjo 
@joaomoreno 
I think my PR #24860 addresses this albeit without a specific shortcut.
it allows The Tab Key to be used to get to the ad and stage icons on the focused line/file
without having to press Enter to select the line which would open the diff in the
editor.

**Comment by yetithefoot** [comment]
I came from Sublime Text. I'm using [GitSavvy](https://github.com/divmain/GitSavvy#highlights) plugin for ST, which implements great user experience working with git via shortcuts.  It would be possible to implement similar experience via VS Code keybindings, but I can find specific actions and options to specify `when` scope for scm viewlet. Appreciate any help.

https://cloud.githubusercontent.com/assets/5016978/6704171/2f236466-cd02-11e4-9b7d-22cc880b5e9d.png

**Comment by franciscolourenco** [comment]
Gitsavvy is by far the best git staging experience I've tried so far. The only thing I'm missing in VSCode atm.

![staging](https://cloud.githubusercontent.com/assets/5016978/6471628/886430f8-c1a1-11e4-99e9-883837dba86f.gif)

Some of the features missing:
- [ ] Open file diff view from directly from editor
- [ ] Ability to stage entire hunks, without having to select them
- [ ] Good Shortcuts, ideally without having to press modifier keys

**Comment by AB1908** [comment]
I similarly came from Atom where you can press Enter on the focused file to Stage/Unstage it. It would be great to have a keyboard shortcut for this. Perhaps we could create a keybinding since the command to stage a file is already in the command palette?

**Comment by LFDM** [comment]
I used an own fork of VSCode for the last to years, where I heavily modified this panel to work in the following way (worth noting that this was also all heavy inspired by vim keybindings):
- When you open the SCM view, the input field is NOT focused, we instead are directly focused at the first unstaged file (I use `alt+g` to open the SCM view.
- `j` and `k` allows to navigate through the list of files
- `a` adds the file to the staged changes, if its already staged it unstages
- `u` undos changes to a file
- `d` diffs the file
- `i` jumps to the input field to write the commit message
- `esc` jumps out of the input field back to the file list.
- `ctrl/cmd+enter` commits it.

`j` would also select the group headers like "Changes" - pressing `a` there would stage all changed files, or unstage all files when you have "Staged Changes" focused.

So the regular flow is for example like this:

`alt+g` to open. `j` `j` `j` to jump to the file I want, `a` to stage the file, `a` to stage the next file, `a` to stage the next file, `i` to start writing the commit message, `ctrl+enter` to commit.

It's blazingly quick and super convenient. Definitely something I am missing from regular VSCode.

**Comment by kgrossjo** [comment]
It turns out that quite a few things have changed, usability is much improved.  It's still possible to do better, though.

Say I'm in a project that has a git repo and some files are changed ("unstaged changes" in `git status`).

`ctrl-shift-g` opens the SCM view, the commit message field is focused.  I can do `tab` to go to the list of files ("Changes"), then `j` to the file I want, then `tab tab tab` to focus the "+" icon, then `space` to add the file to the list of "Staged Changes".

But the SCM view has lost focus, so I have to `ctrl-shift-g` again, which focuses the commit message field.  From there, I can `tab` to the list of files, then `j` my way down over all the files I've already staged down to the unstaged changes, to the file I want.

I feel that it would be nice if, after hitting `space` to execute the "+" action to move the file to the list of staged changes, the SCM view retained focus, and perhaps the first unstaged file could have focus, or one file down from the one I "+"-ed, or one file up from the one I "+"-ed.

I also like the approach presented by @LFDM in his comment from a few hours ago.  That after `ctrl-shift-g`, the list of files has focus, and that there is a key to move focus to the commit message.

**Comment by simkimsia** [comment]
I tried this but didn't work as well.

![image](https://user-images.githubusercontent.com/245021/60566172-b04df680-9d98-11e9-8751-715471be57d7.png)


WOuld be good if there's a clear way on how to set the when expression for scm view is in focus

**Comment by electrovir** [comment]
Is this dead? :( Reviewing commits that touch many files in VS Code is a very slow experience right now...

**Comment by CombeeMike** [comment]
Since my feature request #50434 has been closed as a duplicate of #24389 which then has been merged with this one, I'd like to stress here that a dedicated command (with shortcut) to open the next file with changes in the currently open diff view would improve my VCS/GIT workflow tremendously (as described in #50434)!

Some additional thoughts:

* If no diff view is open, show the Source Control panel & open a new diff editor with the first file with changes
* If there are no more files with changes (the last one in the list is reached), either show the first one again or show some according message/toast to the user. I'd prefer the second option since it makes it a bit clearer that I've reviewed all my changes.

# Workaround using AHK

In the meantime I managed to workaround this by using some AHK script which is bound to a global Windows hotkey:

```
If WinActive("ahk_exe Code.exe")
{
  Send ^+m     ; Ctrl+Shift+m -> Open "Source Control" panel and focus commit message input box (VSCode command "workbench.view.scm")
  Send {Tab}   ; Focus list of changed files
  Send {Down}  ; Select next file in list
  Send {Enter} ; Focus diff editor
}
Else
{
  MsgBox, Not in VSCode!
}
```

This works pretty well so far but there's one thing missing for me: I haven't found a way to automatically select the first item in the list using AHK.

# Question

Does anybody know how to select the first item in the list using the keyboard without knowing how many items there are in the list & which item is currently selected?

In other words: When the list has focus, is there a keyboard command which always selects the first item in the list?

### PS

I hope I'm not "hijacking" this thread too much here...

**Comment by johndevedu** [comment]
This has been bugging me for a while. Finally came up w/ a solution that works pretty well. **TL:DR instead of navigating through each file one by one in the SCM panel, get all files open (requires GitLens extension) then review one file at a time -> close tab when done.**

1. Close all tabs (optional, but cleaner experience for me)
2. Use GitLens: Open All Changes (with difftool). I have difftool connected to VSCode.
{
  "key": "alt+numpad6",
  "command": "gitlens.externalDiffAll"
}
3. Show Next Change / Show Previous Change
{
  "key": "alt+numpad3",
  "command": "workbench.action.compareEditor.nextChange",
  "when": "textCompareEditorVisible"
}
4. Once done reviewing each file, close tab
5. Repeat steps 1-4 for other files until all files complete.

**Comment by PurplProto** [comment]
Sadly this still seems like quite a pain point for many 😢. 

Not a perfect solution but the current way I manage to keep my hands on the keyboard without any fancy extensions or forked Code builds is rather long-winded although, it works-ish:

1. Focus the SCM panel using `Ctrl+Shift+G G` (I believe Git Lens changes this from the default)
2. Use the arrow keys to highlight a change in the list
2. Press `Space` to open the diff, or `Enter` if I want to focus the editor to scroll the file or use `Alt+F5`
3. When happy with reviewing the changes, use `Ctrl+Shift+G G` to focus the SCM panel again
4. Press `Tab` to focus the actions section on the file
5. Press `LeftArrow` to highlight the stage button
6. Press `Space` to stage the change
7. Repeat 2-6 as many times as required
8. Press `PgUp` to highlight the commit message box
9. Press either `Space` or `Enter` to focus the message box
10. Type your commit message
11. Press `Ctrl+Enter` to commit

Note, from step 9 you can press `Shift+Tab` to lose focus of the commit message box and return to having the SCM panel focused so you can stage/unstage files again.

**Comment by dannypernik** [comment]
Can we call a vote on this one?

**Comment by AB1908** [comment]
Apologies in advance for not contributing meaningfully but I've since shifted to vim for writing commit messages since the terminal is a simple enough binding away. Hope y'all find a resolution though!

**Comment by wjandrea** [comment]
> I tried this but didn't work as well.
> 
> ![image](https://user-images.githubusercontent.com/245021/60566172-b04df680-9d98-11e9-8751-715471be57d7.png)
> 
> WOuld be good if there's a clear way on how to set the when expression for scm view is in focus

@simkimsia `git.stageChange` doesn't seem to do anything. It's not available in the command palette when focused on the SCM, editor, or a diff, so I'm not sure how it's supposed to be used. Maybe it's available in some other context?

But anyway, I think you want to use `focusedView == "workbench.scm"` instead of `activeViewlet == "workbench.view.scm"` so that the shortcut only works when the SCM is _focused_, not just when it's _visible as a viewlet_. (Personally, I actually pinned mine to the other side (like in Atom), which makes it not a viewlet.)

**Comment by Tantol** [comment]
> I used an own fork of VSCode for the last to years, where I heavily modified this panel to work in the following way (worth noting that this was also all heavy inspired by vim keybindings):
> 
>     * When you open the SCM view, the input field is NOT focused, we instead are directly focused at the first unstaged file (I use `alt+g` to open the SCM view.
> 
>     * `j` and `k` allows to navigate through the list of files
> 
>     * `a` adds the file to the staged changes, if its already staged it unstages
> 
>     * `u` undos changes to a file
> 
>     * `d` diffs the file
> 
>     * `i` jumps to the input field to write the commit message
> 
>     * `esc` jumps out of the input field back to the file list.
> 
>     * `ctrl/cmd+enter` commits it.
> 
> 
> `j` would also select the group headers like "Changes" - pressing `a` there would stage all changed files, or unstage all files when you have "Staged Changes" focused.
> 
> So the regular flow is for example like this:
> 
> `alt+g` to open. `j` `j` `j` to jump to the file I want, `a` to stage the file, `a` to stage the next file, `a` to stage the next file, `i` to start writing the commit message, `ctrl+enter` to commit.
> 
> It's blazingly quick and super convenient. Definitely something I am missing from regular VSCode.

This functionality is what I need to replace lazygit

**Comment by mawid6** [comment]
> I tried this but didn't work as well.
> 
> ![image](https://user-images.githubusercontent.com/245021/60566172-b04df680-9d98-11e9-8751-715471be57d7.png)
> 
> WOuld be good if there's a clear way on how to set the when expression for scm view is in focus

Yeah, this feels like a bug to me. 

I dont understand how to set the stage and unstage actions to be executed, if possible. I also have tried `focusedView == "workbench.scm"`, and other combinations with different key bindings (for me, plus and minus would seem logical, for instance), and two things to notice:

1. The bindings show up in the context menu! As if they should be working! See this screenshot:

![image](https://github.com/microsoft/vscode/assets/77722885/dc276d93-91af-4fe7-9ea1-481fe3e3f1c5)

2. Using the binding, I have intentionally used ctrl+-, which "zooms out", but if I am in the source control panel with this binding defined, it does not zoom out. This seems to me that the binding actually works, it catches the keypress combo and likely tries to start the command but something is amiss with the unstage command - it might be expecting some other context, or whatever. But something is hindering unstage from executing.


## So, it feels like this specific situation / keybinding not working is actually a bug.
What do you reckon?

**Comment by wjandrea** [comment]
@mawid6 Use the command **Developer: Toggle Keyboard Shortcuts Troubleshooting** to see what's really going on

**Comment by mawid6** [comment]
Thanks alot for the tip @wjandrea! I was not aware of that -so far, I am staying away from vs code development... :-) 
However, the log helped me further pin down the problem and confirm my hypothesis above:
`git.stage` command _is_ in fact being invoked, but it does not handle the event properly, but instead fails silently.

I will update my comment above, transforming it from a hypothesis to a statement and a further hypothesis on what is wrong.

**Comment by dannypernik** [comment]
@mawid6 do you think it would be worth creating a new issue that references this one in order to bring more attention to it?This issue is much broader in scope than what you've identified. We're approaching 7 years in the making here.

**Comment by mawid6** [comment]
@dannypernik well, to be fair, it seems like they could solve the OP problem if they fixed so that we can access the SCM commands (whereof stage was in focus for the OP) from the panel. However, I think I might create a new issue to bring attention to it. That makes sense. Thanks.

**Comment by starball5** [comment]
Related on Stack Overflow: https://stackoverflow.com/q/78856783/11107541

**Comment by FloppyDisco** [comment]
The git.stage command is definitely what is being run when you click on the "+" icon to stage the file. I would suspect, when the command is invoked from the keybinding it does not have access to the active_item from the list of changes, and thus does not know which file to stage. 

this is a pretty silly bug to have not been fixed for so long. 

I made an extension to stage files with the keyboard as a workaround, hope it helps.

https://marketplace.visualstudio.com/items?itemName=bs-code.git-quick-stage

**Comment by b1mind** [comment]
> 
> this is a pretty silly bug to have not been fixed for so long.
> 

Seriously... keyboard accessibility sounds like it's on the priority list these days at VSC. This one drives me up the wall though.... FIX IT PLEASE!! 

I had to come up with a customMultiCommand bind just to be able to stage one file with a keypress and it's still annoying. Would love to be able to simply have it work and work well with multiple files selected... like how the clicking + stage button works. 

I might try out your extension @FloppyDisco appreciate the effort.

**Comment by dannypernik** [comment]
@FloppyDisco your extension is a great start! It appears that new files fail to stage:
![Image](https://github.com/user-attachments/assets/fb36f1cb-f595-44d3-bd87-14447f9c6419)

Hey, maybe we can get some action taken on this issue thread after all! Thank you for your efforts!

**Comment by The-Compiler** [comment]
You might also want to look at [Edamagit](https://github.com/kahole/edamagit), which reimplements a lot of git functionality inside an editor window for a fully keyboard driven interface. You can easily stage files or even individual hunks that way:

![demo](https://github.com/kahole/edamagit/raw/5ebf38107c6130cc16a23f18d84aeecc21f09fe8/magit_commit_demo.gif)

**Comment by franciscolourenco** [comment]
@The-Compiler I've tried Edamagit, unfortunately it is not at the same level as GitSavvy.

**Comment by dannypernik** [comment]
> However, I think I might create a new issue to bring attention to it. That makes sense. Thanks.

@mawid6 curious if you got around to this. If so, can you link the issue here?

**Comment by FloppyDisco** [comment]
> [@FloppyDisco](https://github.com/FloppyDisco) your extension is a great start! It appears that new files fail to stage: 

hey i created an issue, [here](https://github.com/FloppyDisco/git-stage-file/issues/4) '#4' tell me more about it. it seems to stage new files for me just fine.
